### ISSUE TITLE
Support additional Refaster features seen in error-prone-support

### ISSUE THREAD

**Issue by timtebeek** [openrewrite/rewrite-templating#47]
## What problem are you trying to solve?
Expand the set of Refaster rules in Error Prone Support that we cover with OpenRewrite recipes.

## Describe the solution you'd like
Support the following cases not currently covered.
- [ ] Generics as seen in [ImmutableSortedMapRules.java](https://github.com/PicnicSupermarket/error-prone-support/blob/master/error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ImmutableSortedMapRules.java)
- [ ] `com.google.errorprone.refaster.Refaster` as seen in [EqualityRules.java](https://github.com/PicnicSupermarket/error-prone-support/blob/master/error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/EqualityRules.java)
  - #21 
- [ ] `@UseImportPolicy` as seen in [MockitoRules.java](https://github.com/PicnicSupermarket/error-prone-support/blob/master/error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/MockitoRules.java)
- [ ] `@Placeholder` as seen in [LongStreamRules.java](https://github.com/PicnicSupermarket/error-prone-support/blob/master/error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/LongStreamRules.java)
- [ ] `@AlsoNegation` as seen in [BigDecimalRules.java](https://github.com/PicnicSupermarket/error-prone-support/blob/master/error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/BigDecimalRules.java)
- [ ] `@DoNotCall` as seen in [TestNGToAssertJRules.java](https://github.com/PicnicSupermarket/error-prone-support/blob/master/error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/TestNGToAssertJRules.java)
- [ ] `@Repeated` as seen in [WebClientRules.java](https://github.com/PicnicSupermarket/error-prone-support/blob/master/error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/WebClientRules.java)
- [ ] `@Matches` as seen in [ReactorRules.java](https://github.com/PicnicSupermarket/error-prone-support/blob/master/error-prone-contrib/src/main/java/tech/picnic/errorprone/refasterrules/ReactorRules.java)

**Comment by Stephan202** [comment]
Hey @timtebeek! The `@DoNotCall` annotation has no impact for Refaster; it's there only to appease the [`DoNotCall`](https://errorprone.info/bugpattern/DoNotCall) check :)

**Comment by timtebeek** [comment]
> Hey @timtebeek! The `@DoNotCall` annotation has no impact for Refaster; it's there only to appease the [`DoNotCall`](https://errorprone.info/bugpattern/DoNotCall) check :)

Thanks for that context! Right now we skip any rules where that annotation is present, but it seems like we can tolerate the presence of that annotation and just generate the recipes in those cases. Good to know!

**Comment by timtebeek** [comment]
Even if we ignore the presence of `@UseImportOlicy` annotation it seems we still have some work to do before recipe correctly compile, as we don't yet add resolve and imports explicitly. Internally we had relied on `after` templates always using fully qualified refs, but we can't expect that going forward.

**Comment by knutwannheden** [comment]
@timtebeek I think we can. I thought the template code is already being generated with all fully qualified references (regardless if the input uses qualified references or not). That was in any case the idea.

**Comment by timtebeek** [comment]
> @timtebeek I think we can. I thought the template code is already being generated with all fully qualified references (regardless if the input uses qualified references or not). That was in any case the idea.

Exploring the addition of new static imports now in this PR; based on what we do versus need (https://github.com/openrewrite/rewrite-templating/pull/53/commits/d1ce2f70a67ae270d6778ca439102d4d0ddb2a41) it looks like we have some work to do still.
- https://github.com/openrewrite/rewrite-templating/pull/53

**Comment by timtebeek** [comment]
Some quick stats on why various refaster rules were excluded, such that we know what to cover next:
```
[INFO] 791x Generics are currently not supported
[INFO] 40x com.google.errorprone.refaster.Refaster is currently not supported
[INFO] 19x @AlsoNegation is currently not supported
[INFO] 15x @Placeholder is currently not supported
[INFO] 8x If statements are currently not supported
[INFO] 1x @Matches is currently not supported
```

**Comment by knutwannheden** [comment]
Basic generics support is available. Not yet for gwneric type variables, however.

**Comment by timtebeek** [comment]
With Knut's support for generics and a very minor new exclusion, these are the new stats with regards to what's supported/skipped:
- https://github.com/openrewrite/rewrite-templating/pull/72
```
[INFO] @Matches is currently not supported=1
[INFO] @Repeated is currently not supported=1
[INFO] Method references are currently not supported=1
[INFO] Lambdas are currently not supported=2
[INFO] If statements are currently not supported=8
[INFO] @AlsoNegation is currently not supported=11
[INFO] @Placeholder is currently not supported=15
[INFO] com.google.errorprone.refaster.Refaster is currently not supported=75
[INFO] Generic type parameters are currently not supported=495
```

**Comment by timtebeek** [comment]
As discussed yesterday with @rickie we might want to detect any use of OnlineDocumentation, and if present add a link to that URL in the documentation that we generate, both to give credit and additional context to what a recipe would effectively do.
https://github.com/PicnicSupermarket/error-prone-support/blob/8f64489fa0dded9255d20c227d0810af1cf52493/refaster-support/src/main/java/tech/picnic/errorprone/refaster/annotation/OnlineDocumentation.java#L44-L48

We already parse any JavaDoc that's present (even if it's often missing); we might want to expand (or generate) the Javadocs present in Error Prone Support to improve how those are presented to our users.

**Comment by rickie** [comment]
About the `@OnlineDocumentation`, it has default values that are used by Picnic. However, others can use the annotation to link to their own website if they have one. So supporting the annotation will not be a "Picnic-specific" solution.

**Comment by timtebeek** [comment]
We can also already generate templates for any rules that use `@AlsoNegation`, without generating that negation just yet. Adding that negated recipe can then be left as a next step.

**Comment by greg-at-moderne** [comment]
> Some quick stats on why various refaster rules were excluded, such that we know what to cover next:

Current status update on this is:
```
@Matches is currently not supported=1
com.google.errorprone.refaster.Refaster is currently not supported=1
@NotMatches is currently not supported=2
Lambdas are currently not supported=2
Method references are currently not supported=3
If statements are currently not supported=8
@Placeholder is currently not supported=15
@Repeated is currently not supported=21
Generic type parameters are currently not supported=511
```

=== 
For prosperity this has been generated with:

1. Enabling the lines at: https://github.com/openrewrite/rewrite-templating/blob/bb9015732d52cc59077667106261d849c9f9d149/src/main/java/org/openrewrite/java/template/processor/RefasterTemplateProcessor.java#L117-L119
1. `./gradlew pTML`
1. publishing a local snapshot of `rewrite-templating` with `publishToMavenLocal`
1. check out error-prone-support
1. take the version string and plug into https://github.com/PicnicSupermarket/error-prone-support/blob/cf294b8964ae7e67c0014de938b04b1023417e82/pom.xml#L218
1. `cd error-prone-contrib; mvn clean compile` and carefully look at the output

**Comment by knutwannheden** [comment]
Not to be forgotten here are the `@AlsoNegation` annotations, which we don't yet support. We currently just generate the "positive" recipes.

**Comment by timtebeek** [comment]
> Not to be forgotten here are the `@AlsoNegation` annotations, which we don't yet support. We currently just generate the "positive" recipes.

Indeed; these are used in ~9 Rules files in error-prone-contrib: https://github.com/search?q=repo%3APicnicSupermarket%2Ferror-prone-support%20AlsoNegation&type=code . Would be nice to cover but the big win is in adding support for generics.

**Comment by greg-at-moderne** [comment]
Continually updated list of PRs towards providing support for generic type parameters:
- https://github.com/openrewrite/rewrite/pull/5058
- ... more to come

**Comment by timtebeek** [comment]
This will go a long way to closing the gap; curious to try it out:
- https://github.com/openrewrite/rewrite-templating/pull/134

**Comment by greg-at-moderne** [comment]
> Some quick stats on why various refaster rules were excluded, such that we know what to cover next:

Using the same methodology as above, one can arrive at:
```
@NotMatches is currently not supported=7
Lambdas are currently not supported=7
If statements are currently not supported=10
com.google.errorprone.refaster.Refaster is currently not supported=13
Refaster.anyOf() can only be used once per template=15
Method references are currently not supported=16
@Matches is currently not supported=30
@Placeholder is currently not supported=54
@Repeated is currently not supported=59
```
